File: 剑指offer/29. 顺时针打印矩阵.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix: return []
        t, r, b, l, res = 0, len(matrix[0])-1, len(matrix)-1, 0, []
        while True:
            for i in range(l, r+1): res.append(matrix[t][i])
            t += 1
            if t>b: break              # t>b错写成t<b,记住t的初始下标是0，越往下下标越大
            for i in range(t, b+1): res.append(matrix[i][r])
            r -= 1
            if l>r: break
            for i in range(r, l-1, -1): res.append(matrix[b][i]) # range(r, l-1, -1)忘记下标是一路递减的，漏写了
            b -= 1
            if t>b: break
            for i in range(b, t-1, -1): res.append(matrix[i][l])
            l += 1
            if l>r: break
        return res
    
#时间复杂度: O(M*N)
#空间复杂度: O(1)
https://leetcode-cn.com/problems/shun-shi-zhen-da-yin-ju-zhen-lcof/solution/mian-shi-ti-29-shun-shi-zhen-da-yin-ju-zhen-she-di/

    
    
    

File: 剑指offer/C++_Version/29. 顺时针打印矩阵.cpp
//方法1
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix)
    {   
        vector<int> res;
        if(matrix.empty()) return res;
        int t=0, r=matrix[0].size()-1, b=matrix.size()-1, l=0;
        while(1==1)
        {
            for(int i=l; i<r+1; i++) res.push_back(matrix[t][i]);
            t++;
            if(t > b) break;
            for(int i=t; i<b+1; i++) res.push_back(matrix[i][r]);
            r--;
            if(l > r) break;
            for(int i=r; i>l-1; i--) res.push_back(matrix[b][i]);
            b--;
            if(t > b) break;
            for(int i=b; i>t-1; i--) res.push_back(matrix[i][l]);
            l++;
            if(l > r) break;
        }
        return res;
    }
};
